python/pyarrow/_dataset.pyx

```diff
@@ -41,6 +42,167 @@ def _forbid_instantiation(klass, subclasses_instead=True):
     raise TypeError(msg)
 
 
+cdef class Expression:
+
+    cdef:
+        shared_ptr[CExpression] wrapped
+        CExpression* expr
+
+    def __init__(self, Buffer buffer=None):
+        if buffer is not None:
+            c_buffer = pyarrow_unwrap_buffer(buffer)
+            expr = GetResultValue(CExpression.Deserialize(deref(c_buffer)))
+            self.init(expr)
```

Comment:
This would definitely be better but it gives me an error like 
```
E           _pickle.PicklingError: Can't pickle <built-in function _deserialize>: attribute lookup _deserialize on pyarrow._dataset failed
```
I'll permute until this works.